server/leaderboard.js

```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
+                y = b.points * ((b.correctAttemptsCount / b.totalAttemptsCount) + (b.points / b.totalAttemptsCount));                
+                
+                if (a.totalAttemptsCount === 0 && b.totalAttemptsCount ===0){
+                    return 0;
+                } else if (b.totalAttemptsCount === 0){
+                    return -1;
+                } else if (a.totalAttemptsCount === 0){
+                    return 1;
+                } else{
+                    return y - x;
+                }
+            })
+        } 
+        
         /* assign ranks to each student */
         for (var i = 0; i < studentlist.length; ++i) {
             /* subsequent users with same amount of points have same rank */
-            if (studentlist[i].points != last) {
-                rank = i + 1;
-                last = studentlist[i].points;
-            }
-            studentlist[i].rank = rank;
+            
+            //Rank is assigned depending on the type
+
+            if(type === 'points'){
+                if (studentlist[i].points != last) {
+                    rank = i + 1;
+                    last = studentlist[i].points;
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
+                    userind = i;
+
+            } else if(type === 'attempt'){
+                var attempt = 0;
+                if(studentlist[i].totalAttemptsCount != 0){
+                    attempt = studentlist[i].points/studentlist[i].totalAttemptsCount;
+                    attempt = Math.round(attempt);
+                }
 
-            if (studentlist[i].id == userid)
-                userind = i;
+                if (attempt != last) {
+                    rank = i + 1;
+                    last = attempt;
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
+                    userind = i;
+
+            } else if(type === 'accuracy'){
+                var accuracy = 0;
+                if (studentlist[i].totalAttemptsCount != 0){
+                    accuracy = studentlist[i].correctAttemptsCount/studentlist[i].totalAttemptsCount;
+                    accuracy = Math.round(accuracy*100)/100;
+                }
+
+                if (accuracy != last) {
+                    rank = i + 1;
+                    last = accuracy;
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
+                    userind = i;
+
+            } else if(type === 'overall'){
+                var overall = 0;
+                if (studentlist[i].totalAttemptsCount != 0){
+                    overall = studentlist[i].points * ((studentlist[i].correctAttemptsCount / studentlist[i].totalAttemptsCount) + (studentlist[i].points / studentlist[i].totalAttemptsCount));
```

Comment:
very confusing, this should be broken up to individual pieces

Reply:
I'll break it up.